### ISSUE TITLE
Rule proposal: no duplicate type annotations on functions

### ISSUE THREAD

**Issue by OliverJAsh** [typescript-eslint/typescript-eslint#5658]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

No duplicate type annotations on functions.

The fixer could be configured to prefer a function annotation or inline annotations on the function itself.

### Fail Cases

```typescript
const f: (n: number) => number = (n: number) => n;

const f: (n: number) => number = (n: number): number => n;
```

### Pass Cases

```typescript
const f: (n: number) => number = (n) => n;

const f = (n: number): number => n;

const f = (n: number) => n;
```

**Comment by JoshuaKGoldberg** [comment]
I like this a lot. @bradzacher has previously had a lot of thoughts on function type annotations, so I'll defer to Brad.

**Comment by OliverJAsh** [comment]
To make sure we're all on the same page, this is what I meant by "function annotation" and "inline annotations".

Function annotation:

```ts
const f: (n: number) => number = n => n
```

Inline annotations for parameters and the return value, inside the function definition:

```ts
const f = (n: number): number => n;
```

> Similarly how would the rule "fix" this case if it was in "ban inline" mode:

> I guess the same question should be asked about non-nested functions first.

This rule isn't about preferring one style over the other. It's only concerned with avoiding duplication. Therefore these examples would never error—there's no duplication. All of the types are defined via a function annotation and none of the types are defined via inline annotations.

```ts
const f1: (n: number) => (n: number) => number = n => n => n;

const f1: (n: number) => number = n => n;
```

> How does the rule handle nested functions. For example:

> Also would the rule deal with body-less arrows only, or would it deal with bodied arrows and fn expressions:

Good question. I don't know 😄

**Comment by OliverJAsh** [comment]
> How does the rule handle nested functions. For example:

I think this rule should only check for duplicate type annotations between direct pairs, where a function has both a function type annotation and inline type annotations:

```ts
const f2: (n: number) => (n: number) => number = (n): (n: number) => number => n => n;
```

This would error because the function definition `(n): (n: number) => number => n => n` has an inline return type annotations, which duplicates information defined in the function annotation.

```ts
const f3: (n: number) => (n: number) => number = n => (n): number => n;
```

This would not error because the function definition `n => (n): number => n` has no inline type annotations. It's return value does, but _this_ function does not.

**Comment by JoshuaKGoldberg** [comment]
Accepting PRs, but note that there are unresolved discussion points. 

> How would the rule handle ...
> would it fix to ...

I'd suggest adding options for each of those things. My prediction is that this will be a rule that triggers a lot of people's personal opinions for how to declare types.

**Comment by JoshuaKGoldberg** [comment]
Good news! We're planning on having a bigger rule check for unnecessary type annotations as a whole: #12088. This issue's suggested rule is but a part of that one. Closing this issue out as a duplicate - it helped inspire discussion in #12088. Cheers!
